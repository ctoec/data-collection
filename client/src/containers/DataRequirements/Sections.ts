// TO DO: also get this from the backend
export const SECTIONS = {
  CHILD_INFO: 'child-info',
  FAMILY_INFO: 'family-info',
  FAMILY_INCOME: 'family-income',
  ENROLLMENT_FUNDING: 'enrollment-funding',
};

type SectionInfo = { formattedName: string; description: string };
export const SECTION_COPY: { [key: string]: SectionInfo } = {};
SECTION_COPY[SECTIONS.CHILD_INFO] = {
  formattedName: 'Child info',
  description: 'A unique person enrolled in an ECE program.',
};
SECTION_COPY[SECTIONS.FAMILY_INFO] = {
  formattedName: 'Family info',
  description:
    'One or more children that share the same address and household income.',
};
SECTION_COPY[SECTIONS.FAMILY_INCOME] = {
  formattedName: 'Family income determination',
  description:
    "A determination by a provider of a family's income, for purposes of assessing eligibility for public funding; must be updated at least once a year.",
};
SECTION_COPY[SECTIONS.ENROLLMENT_FUNDING] = {
  formattedName: 'Enrollment and funding',
  description:
    'Enrollment: A period of time during which a child recieved ECE services.\nFunding: A period of time during which an enrollment was subsidized by a state-funded contract space.\nCare 4 Kids: Whether or not the enrollment was subsidized by the Care 4 Kids program.',
};
