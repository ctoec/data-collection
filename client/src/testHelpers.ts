import React from 'react';
import { RenderResult, render, cleanup, act } from '@testing-library/react';
import { axe } from 'jest-axe';

export const accessibilityTestHelper = (
  component: React.ReactElement,
  before?: (_: RenderResult) => Promise<void>
) => {
  it('passes AXE accessibility checks', async () => {
    await act(async () => {
      const renderResult = render(component);
      if (before) {
        await before(renderResult);
      }
      const results = await axe(renderResult.container);
      expect(results).toHaveNoViolations();
    });
    cleanup();
  });
};

export const snapshotTestHelper = (component: React.ReactElement) => {
  it('matches snapshot', async () => {
    await act(async () => {
      const { asFragment } = render(component);
      expect(asFragment()).toMatchSnapshot();
    });
  });
};
