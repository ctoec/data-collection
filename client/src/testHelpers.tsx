import React from 'react';
import { RenderResult, render, cleanup, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import { BrowserRouter } from 'react-router-dom';

export const renderHelper = async (
  component: React.ReactElement,
  opts: {
    before?: (_: RenderResult) => Promise<void>;
    wrapInRouter?: boolean;
  } = {}
) => {
  const renderResult = render(
    opts.wrapInRouter ? <BrowserRouter>{component}</BrowserRouter> : component
  );

  if (opts.before) {
    await opts.before(renderResult);
  }
  return renderResult;
};

export const accessibilityTestHelper = (
  component: React.ReactElement,
  opts: {
    before?: (_: RenderResult) => Promise<void>;
    wrapInRouter?: boolean;
  } = {}
) => {
  it('passes AXE accessibility checks', async () => {
    await act(async () => {
      const renderResult = await renderHelper(component, opts);
      const results = await axe(renderResult.container);
      expect(results).toHaveNoViolations();
    });

    cleanup();
  });
};

export const snapshotTestHelper = (
  component: React.ReactElement,
  opts: {
    before?: (_: RenderResult) => Promise<void>;
    wrapInRouter?: boolean;
  } = {}
) => {
  it('matches snapshot', async () => {
    await act(async () => {
      const renderResult = await renderHelper(component, opts);
      expect(renderResult.asFragment()).toMatchSnapshot();
    });

    cleanup();
  });
};
