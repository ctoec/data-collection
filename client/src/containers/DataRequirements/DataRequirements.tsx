import React, { useState, useContext, useEffect } from 'react';
import ReactMarkdown from 'react-markdown';
import { ReactComponent as ArrowLeft } from 'uswds/dist/img/arrow-left.svg';
import { ColumnMetadata } from 'shared/models';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import { Table, Column, TextWithIcon, Button } from '@ctoec/component-library';
import { SECTION_COPY } from './Sections';
import { apiGet } from '../../utils/api';

const DataRequirements: React.FC = () => {
  const { accessToken } = useContext(AuthenticationContext);
  const [columnMetadata, setColumnMetadata] = useState<ColumnMetadata[]>(
    []
  );

  useEffect(() => {
    apiGet('column-metadata').then((definitions) =>
      setColumnMetadata(definitions)
    );
  }, [accessToken]);

  const columns: Column<ColumnMetadata>[] = [
    {
      name: 'Field name',
      cell: ({ row }) =>
        row ? (
          <th scope="row">
            <span className="text-bold">{row.formattedName}</span>
          </th>
        ) : (
          <></>
        ),
    },
    {
      name: 'Required/ Optional',
      cell: ({ row }) => (row ? <td> {row.required} </td> : <></>),
    },
    {
      name: 'Definition',
      cell: ({ row }) =>
        row ? (
          <td>
            <ReactMarkdown source={row.definition} />
          </td>
        ) : (
          <></>
        ),
    },
    {
      name: 'Reason for collecting',
      cell: ({ row }) => (row ? <td> {row.reason} </td> : <></>),
    },
    {
      name: 'Format',
      cell: ({ row }) =>
        row ? (
          <td>
            <div>{row.format}</div>
            <div className="margin-top-1">Ex: {row.example}</div>
          </td>
        ) : (
          <></>
        ),
    },
  ];

  const columnMetadataBySection: { [key: string]: ColumnMetadata[] } = {};
  columnMetadata.reduce((acc, cur) => {
    if (acc[cur.section]) {
      acc[cur.section].push(cur);
    } else {
      acc[cur.section] = [cur];
    }

    return acc;
  }, columnMetadataBySection);

  return (
    <div className="grid-container margin-top-4">
      <Button
        className="text-bold margin-bottom-4"
        appearance="unstyled"
        href="/"
        text={<TextWithIcon text="Back" Icon={ArrowLeft} iconSide="left" />}
      />
      <h1>OEC's enrollment data requirements</h1>
      {Object.entries(columnMetadataBySection).map(
        ([sectionKey, sectionData]) => (
          <div className="margin-top-4">
            <h2>{SECTION_COPY[sectionKey].formattedName}</h2>
            <p className="text-pre-line">
              {SECTION_COPY[sectionKey].description}
            </p>
            <Table
              id="data-requirements-table"
              data={sectionData}
              rowKey={(row) => (row ? row.formattedName : '')}
              columns={columns}
              defaultSortColumn={0}
            />
          </div>
        )
      )}
    </div>
  );
};

export default DataRequirements;
