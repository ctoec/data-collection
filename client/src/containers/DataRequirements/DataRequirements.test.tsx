import React from 'react';
import DataRequirements from './DataRequirements';
import { accessibilityTestHelper, snapshotTestHelper } from '../../testHelpers';
import { BrowserRouter } from 'react-router-dom';

const routerWrapped = (
  <BrowserRouter>
    <DataRequirements />
  </BrowserRouter>
);
describe('DataRequirements', () => {
  snapshotTestHelper(routerWrapped);
  accessibilityTestHelper(routerWrapped);
});
