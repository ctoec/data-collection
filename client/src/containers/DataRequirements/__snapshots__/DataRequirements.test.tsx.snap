// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataRequirements matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container"
  >
    <button
      class="usa-button usa-button--unstyled margin-top-2 text-bold"
      type="button"
    >
      <span
        class="oec-text-with-icon"
      >
        <svg
          class="oec-text-with-icon__icon oec-text-with-icon__icon--direction-left oec-text-with-icon__icon--side-left"
        >
          arrowRight.svg
        </svg>
        Back
      </span>
    </button>
    <h1
      class="margin-bottom-0"
    >
      OEC's enrollment data requirements
    </h1>
    <p
      class="margin-top-05"
    >
      Last updated: 
      <span
        class="text-bold"
      >
        June 21, 2008
      </span>
    </p>
    <p>
      ECE Reporter will allow you to share your state-funded enrollment data with the Office of Early Childhood (OEC).
    </p>
    <p>
      <span
        class="text-bold"
      >
        Enrollments to include
      </span>
      <br />
      Children receiving funding through the following programs:
    </p>
    <ul>
      <li>
         Child Day Care
      </li>
      <li>
         Competitive School Readiness
      </li>
      <li>
         Priority School Readiness
      </li>
      <li>
         State Head Start
      </li>
      <li>
         Smart Start
      </li>
    </ul>
    <p>
      <span
        class="text-bold"
      >
        Data collection period
      </span>
      <br />
      The data collection period begins 07/01/2020 and continues through 12/31/2020.
    </p>
    <p>
      <span
        class="text-bold"
      >
        Data submission deadline
      </span>
      <br />
      You must submit your data to OEC by 03/05/2021.
    </p>
    <h2
      class="border-top border-base-lighter padding-top-1"
    >
      Data elements
    </h2>
    <div
      class="oec-loading-wrapper"
    >
      <div
        aria-hidden="false"
        class=""
      >
        <div
          class="data-definitions"
        >
          <table>
            <caption
              class="text-bold text-left margin-bottom-1"
            >
              Requirement levels
            </caption>
            <tbody>
              <tr>
                <td>
                  <span
                    class="usa-tag usa-tag--no-transform required-tag required-tag--required"
                  >
                    Required
                  </span>
                </td>
                <td>
                  Must be entered for all enrollments
                </td>
              </tr>
              <tr>
                <td>
                  <span
                    class="usa-tag usa-tag--no-transform required-tag required-tag--conditional"
                  >
                    Required If
                  </span>
                </td>
                <td>
                  Must be entered for certain scenarios (such as for an exited enrollment)
                </td>
              </tr>
              <tr>
                <td>
                  <span
                    class="usa-tag usa-tag--no-transform required-tag"
                  >
                    Optional
                  </span>
                </td>
                <td>
                  Suggested for better data quality and record matching, but not required for submission to OEC
                </td>
              </tr>
            </tbody>
          </table>
          <fieldset
            aria-labelledby="fieldset-legend-filter-fields"
            class="usa-fieldset requirement-level-filter"
            id="filter-fields"
          >
            <legend
              id="fieldset-legend-filter-fields"
            >
              <span
                class="usa-label"
              >
                Filter fields
              </span>
            </legend>
            <div>
              <span>
                <div
                  class="usa-radio"
                >
                  <input
                    checked=""
                    class="usa-radio__input"
                    id="all"
                    name="field filter option"
                    type="radio"
                    value="all"
                  />
                  <label
                    class="usa-radio__label"
                    for="all"
                  >
                    All fields
                  </label>
                </div>
              </span>
              <span>
                <div
                  class="usa-radio"
                >
                  <input
                    class="usa-radio__input"
                    id="required"
                    name="field filter option"
                    type="radio"
                    value="required"
                  />
                  <label
                    class="usa-radio__label"
                    for="required"
                  >
                    Only required  and conditionally required fields
                  </label>
                </div>
              </span>
            </div>
          </fieldset>
          <div>
            <div
              class="margin-top-4"
            >
              <h3>
                Section one
              </h3>
              <p
                class="text-pre-line"
              />
              <table
                class="oec-table"
                id="data-requirements-Section-one"
              >
                <thead>
                  <tr>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-sorted"
                      role="columnheader"
                      scope="col"
                      style="width: 18%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Field name
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Required/ Optional
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 24%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Definition
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Reason for collecting
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Format
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      <span
                        class="text-bold"
                      >
                        Favorite color
                      </span>
                    </th>
                    <td>
                      <span
                        class="usa-tag usa-tag--no-transform required-tag"
                      >
                        Optional
                      </span>
                    </td>
                    <td>
                      <p>
                        The color you like the most
                      </p>
                    </td>
                    <td>
                      So we know your favorite color
                    </td>
                    <td>
                      <p>
                        string
                      </p>
                      <div
                        class="margin-top-1"
                      >
                        Ex: Blue
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              class="margin-top-4"
            >
              <h3>
                Section two
              </h3>
              <p
                class="text-pre-line"
              />
              <table
                class="oec-table"
                id="data-requirements-Section-two"
              >
                <thead>
                  <tr>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-sorted"
                      role="columnheader"
                      scope="col"
                      style="width: 18%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Field name
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Required/ Optional
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 24%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Definition
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Reason for collecting
                      </span>
                    </th>
                    <th
                      aria-sort="none"
                      class="oec-table__column-header oec-unsorted"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="oec-table__column-title"
                      >
                        Format
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      <span
                        class="text-bold"
                      >
                        Lucky number
                      </span>
                    </th>
                    <td>
                      <span
                        class="usa-tag usa-tag--no-transform required-tag required-tag--required"
                      >
                        Required
                      </span>
                    </td>
                    <td>
                      <p>
                        The number you find brings the most luck
                      </p>
                    </td>
                    <td>
                      So we know your lucky number
                    </td>
                    <td>
                      <p>
                        number
                      </p>
                      <div
                        class="margin-top-1"
                      >
                        Ex: 130
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
      <div
        class="oec-loading-wrapper__indicator bg-white"
        hidden=""
      >
        <div
          class="MuiCircularProgress-root indeterminate MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
